### DISCUSSION TITLE
download-youtube-playlists/

### DISCUSSION THREAD

**Discussion by giscus** [amir-pourmand/hugo-papermod-farsi-template#11] [Announcements]
# download-youtube-playlists/

در اینجا روشی برای دانلود پلی‌لیست‌های یوتیوب با کتابخانه Pytube معرفی می‌کنم.

https://aprd.ir/download-youtube-playlists/

**Comment by pourmandmaryam** [comment]
😃!منکه جز دانلود از یوتیوب چیزی نفهمیدم ولی خب جالبه

**Comment by Rohit0266** [comment]
از <a href="https://revancedyoutube.org/">revanced</a> برای حذف تبلیغات در اندروید استفاده کنید. این ادامه‌ی پروژه Youtube Vanced است و به شما امکان می‌دهد تا از تجربه‌ی تماشای ویدئو بدون تبلیغات لذت ببرید.

**Comment by Rohit0266** [comment]
https://revancedyoutube.org/

**Comment by sajadsh7** [comment]
سلام 
اگر ممکنه در مورد اون قسمت تقسیم کردن پلی لیست برای پلی لیست های خیلی بزرگ و بعد دانلود کردن هر کدام به صورت یکSession صحبت کنید

**Reply by pourmand1376** [reply]
سلام سجاد جان. 

من این قسمت رو صرفاً لینک داده بودم به گیت‌هاب و گفتم که اگر لازم شد، بعداً توضیحش می‌دم. همون‌طور که خودت می‌دونی، اگر تعداد فایل‌های پلی‌لیست از یه تعدادی (بستگی به پلی‌لیست داره) بیشتر باشه، نمیشه یه باره دانلودشون کرد و لازمه که مثلاً در سه بار جداگانه دانلود بشن. من تو کد [دانلود پلی‌لیست مازیار رئیسی](https://github.com/pourmand1376/Notebooks/blob/main/DownloadMaziar.ipynb) هم همین‌کار رو انجام دادم. 

بیا مرحله به مرحله پیش بریم و کد رو توضیح بدم. اول لینک دانلود ویدئوهای پلی‌لیست رو درمیاریم. 
```python
from pytube import Playlist
playlist = Playlist('https://www.youtube.com/playlist?list=PLoEMreTa9CNmuxQeIKWaz7AVFd_ZeAcy4')
```

خروجی این کد یه ارایه است که داخل playlist ذخیره میشه. بعد من یه کدی نوشتم که بیایم این ارایه رو سه قسمت کنیم و هر کدوم رو داخل یک فایل جداگانه ذخیره کنیم. 
```python
from pathlib import Path
Path('part1.txt').write_text("\n".join(playlist[:150]))
Path('part2.txt').write_text("\n".join(playlist[150:300]))
Path('part3.txt').write_text("\n".join(playlist[300:]))
```
این کد میاد سه تا فایل به نام‌هایی که نوشتم ذخیره می‌کنه. مشخصاً اینجا اومدم تو هر فایل ۱۵۰ تا لینک گذاشتم ولی خب اگر حجم ویدئوهای پلی‌لیستت بیشتر باشه قاعدتاً تعداد کمتری ویدئو می‌تونی رو تو یک بار دانلود کنی. مثلاً ممکنه این عدد برای تو ۵۰ یا ۸۰ یا ۱۰۰ باشه. 

خب الان باید هر کدوم رو جداگانه دانلود کنی و زیپ کنی و روی گوگل درایو قرار بدی. فقط نکته‌اش اینه که این دفعه اینطوری باید ویدئوها رو دانلود کنی. 
```python
! for name in $(cat ../part1.txt);do pytube $name;done
```

کاری که کد بالا انجام میده، سخت نیست. صرفاً میاد یه حلقه تشکیل میده و یکی یکی لینک‌هایی که داخل فایل part1.txt هستند رو دانلود می‌کنه.

بقیه کارهاش مثل روش عادیه؛ یعنی بعد از این که کار تموم شد، فولدرت رو زیپ می‌کنی و روی گوگل درایوت ذخیره می‌کنی. 

قاعدتاً وقتی دانلود هر پارت تموم شد و فایل زیپ‌اش رو دانلود کردی، باید یه فایل جدید Colab باز کنی که اذیتت نکنه. چون ممکنه وسط کار قطعت کنه. 

یه کار دیگه هم میشه انجام داد. بعد از این که لینک‌ها رو تقسیم کردی و تو فایل‌های جداگونه ذخیره کردی، می‌تونی از گزینه New Notebook استفاده کنی و چند تا نوت بوک کولب داشته باشی که همزمان هر کدوم یک قسمتی از این پلی‌لیست عظیم رو برات دانلود کنند. 

فقط فکر کنم تعداد نوت‌بوک‌هایی که می‌تونی همزمان داشته باشی یه محدودیتی داره (احتمالاً نهایتاً ۴ تا) و نمیشه مثلاً ۱۰ تا نوت‌بوک داشته باشی. ولی در همون حد هم کار ما رو راه می‌اندازه چون اینترنت کولب مثل اینترنت ما لاک‌پشتی نیست و سرعتش بسیار بسیار بالاست. پس زیاد معطل نمیشی. 

امیدوارم حرفام برات قابل فهم بوده باشه. بازم اگر جاییش برات مبهم بود بگو بیشتر توضیح بدم.

**Comment by mha27404** [comment]
سلام 
هیج راهی وجود داره بتونیم ویدیو هایی که دانلود میشه به صورت شماره گذاری شده دانلود کنیم .

**Reply by pourmand1376** [reply]
سلام. منظورتون رو متوجه نشدم. میشه واضح‌تر توضیح بدید؟

**Reply by mha27404** [reply]
منظورم اینکه ویدیو ها به همون ترتیبی که داخل پلی لیست هستن شماره گذاری بشه
توی حالت عادی وقتی فایل ها رو دانلود میکنیم هیچ ترتیب خاصی ندارن و به هم ریخته هستن 
من میخوام به همون ترتیبی که حاصب کانال پلی لیست رو مرتب کرده دانلود کنم 
مثلا ما این پلی لیست رو داریم که به این صورت ذخیره شده داخل یوتیوب :
***
1-معرفی هوش مصنوعی برای جابجایی صورت - سایت تغییر چهره با هوش مصنوعی
2-عکس پرنسلی در خانه | آموزش گرفتن عکس پرسنلی در خانه
3-عکس لاتاری را در خانه با گوشی خود بگیرید! آموزش گرفتن عکس لاتاری در خانه
4-آموزش فتوشاپ رایگان/آموزش فتوشاپ: تکنیک ذوب شدن عکس برای اینستاگرام
5-چطور برای ویدیوهای یوتوب در فوتوشاپ، کاور طراحی کنیم/ساخت تامنیل در یوتیوب
6-تبدیل عکس به کارتون در فتوشاپ/نرم افزار تبدیل عکس به نقاشی/کارتونی کردن عکس در فتوشاپ
7-چطور برای ویدیوهای یوتوب در فوتوشاپ، کاور طراحی کنیم-آموزش تامبنیل آریا کئوکسر
8-ادیت عکس در فتوشاپ(تبدیل عکس به نقاشی دیجیتال)
9-آموزش فتوشاپ - چطور رنگ هر چیزی را در فوتوشاپ تغییر بدهیم؟
10-ادیت برش عکس در فتوشاپ
11-آموزش فتوشاپ : افکت نور رنگی در فتوشاپ | نورپردازی عکس در فتوشاپ
***
اما وقتی دانلود میشه چون نام ویدیو ها شماره گذاری نمیشه به این صورت ترتیب بندی میشه:
***
ادیت برش عکس در فتوشاپ
ادیت عکس در فتوشاپ(تبدیل عکس به نقاشی دیجیتال)
آموزش فتوشاپ  افکت نور رنگی در فتوشاپ  نورپردازی عکس در فتوشاپ
آموزش فتوشاپ - چطور رنگ هر چیزی را در فوتوشاپ تغییر بدهیم؟
آموزش فتوشاپ رایگانآموزش فتوشاپ تکنیک ذوب شدن عکس برای اینستاگرام
تبدیل عکس به کارتون در فتوشاپنرم افزار تبدیل عکس به نقاشیکارتونی کردن عکس در فتوشاپ
چطور برای ویدیوهای یوتوب در فوتوشاپ، کاور طراحی کنیم-آموزش تامبنیل آریا کئوکسر
چطور برای ویدیوهای یوتوب در فوتوشاپ، کاور طراحی کنیمساخت تامنیل در یوتیوب
عکس پرنسلی در خانه    آموزش گرفتن عکس پرسنلی در خانه
عکس لاتاری را در خانه با گوشی خود بگیرید! آموزش گرفتن عکس لاتاری در خانه
معرفی هوش مصنوعی برای جابجایی صورت - سایت تغییر چهره با هوش مصنوعی
***

**Reply by pourmand1376** [reply]
همم. تو داکیومنت‌های کتابخانه اگر نگاه کنی، راه حلش هست:
- https://github.com/pytube/pytube/issues/1117

چیزی شبیه این:
```python
from pytube import Playlist, YouTube

playlist_url = 'PLAYLIST_URL_HERE'
p = Playlist(playlist_url)
for index, url in enumerate(p):
    video = YouTube(url)
    video.streams.get_highest_resolution().dowload(filename_prefix=index)
```

**Comment by saravbs** [comment]
سلام
امیدوارم خوب باشید :)

این نوشته رو امروز که داشتم دنیال راهی میگشتم برای دانلود از یوتوب پیدا کردم و به کارم اومد. میخواستم تشکر کنم برای به اشتراک گذاشتن این راه‌حل.
